app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+        val idsToRetrieve: List<String>
+        _hikesMutex.withLock {
+          idsToRetrieve = _hikeFlowsMap.values.filter { !hasOsmData(it.value) }.map { it.value.id }
+        }
+
+        // If all routes already have their OSM data, do nothing more
+        if (idsToRetrieve.isEmpty()) {
+          onSuccess()
+          return@withContext
+        }
+
+        // Retrieve the OSM data of the hikes
+        val hikeRoutes: List<HikeRoute>
+        try {
+          hikeRoutes = loadHikesByIdsRepoWrapper(idsToRetrieve)
+        } catch (e: Exception) {
+          Log.e(LOG_TAG, "Error encountered while loading hikes OSM data", e)
+          onFailure()
+          return@withContext
+        }
+
+        // Update the retrieved data for the loaded hikes
+        _hikesMutex.withLock {
+          hikeRoutes.forEach { hikeRoute ->
+            val hikeFlow = _hikeFlowsMap[hikeRoute.id] ?: return@forEach
+            val hike = hikeFlow.value
+
+            val newHike =
+                hike.copy(
+                    name = hikeRoute.name,
+                    description = DeferredData.Obtained(hikeRoute.description),
+                    bounds = DeferredData.Obtained(hikeRoute.bounds),
+                    waypoints = DeferredData.Obtained(hikeRoute.ways))
+            hikeFlow.value = newHike
+          }
+
+          // Update the selected hike if necessary
+          updateSelectedHike()
+        }
+
+        // Release the mutex before calling onSuccess to avoid deadlocks or performance issues
+        onSuccess()
+      }
+
+  /**
+   * The [HikeRoutesRepository] interface has been developed without coroutines in mind, hence we
+   * need a wrapper to "convert" the [HikeRoutesRepository.getRouteById] function that uses callback
+   * to a suspend function.
+   */
+  private suspend fun loadHikesByIdsRepoWrapper(ids: List<String>): List<HikeRoute> =
+      suspendCoroutine { continuation ->
+        osmHikesRepo.getRoutesByIds(
+            routeIds = ids,
+            onSuccess = { hikeRoutes -> continuation.resume(hikeRoutes) },
+            onFailure = { exception -> continuation.resumeWithException(exception) })
+      }
+
+  private suspend fun retrieveElevationDataForAsync(
+      hikeId: String,
+      onSuccess: () -> Unit,
+      onFailure: () -> Unit
+  ) =
+      withContext(dispatcher) {
+        // Retrieve the hike from the list to perform a few checks
+        var success = false
+        var alreadyComputed = false
+        var mutableWaypoints: List<LatLong>? = null
+        _hikesMutex.withLock {
+          val hikeFlow = _hikeFlowsMap[hikeId] ?: return@withLock
+          val hike = hikeFlow.value
+
+          // If the elevation of this hike is already computed or requested, do nothing
+          alreadyComputed = hike.elevation !is DeferredData.NotRequested
+          if (alreadyComputed) {
+            success = true
+            return@withLock
```

Comment:
Because `onSuccess()` needs to be called (the elevation was already computed or requested), but it needs to be done with the lock on the Mutex, otherwise deadlocks may arise.

Also it lets other tasks waiting on the lock to resume.

Reply:
Like, `alreadyComputed` is only set here and checked below, and below, the only thing you do is check its value, if it's true, then call `onSuccess` and `return@withContext`. 

Why not just here, instead of:
Checking `alreadyComputed`, setting `success` to `true`, `return@Lock`, checking `alreadyComputed` again, calling `onSuccess`, `return@withContext`
Just do here: 
Checking `alreadyComputed`, calling `onSuccess`, `return@withContext`

This shouldn't have any effect on the lock either

Reply:
The issue is if we call `onSuccess` in the lock, we don't know what `onSuccess` contains, it could make another call to the view model and cause a deadlock. Hence, I kept the callbacks out of the lock block.

Reply:
Oooooooh, I see. Thanks for the explanation!